ExprMap produces expressions that are `equals` but have different `hashCode`

## Compiler version

3.1.2, 3.1.3-RC3, 3.2.0-RC1-bin-20220517-e5abec0-NIGHTLY
## Minimized code

The gist of the code is, that it uses an `ExprMap` to collect all subexpressions. It does that twice, and the compares the found expressions.

```Scala
import scala.quoted.*

inline def mac[T](inline expr: T): T =
  ${ impl('expr) }

class MyMap() extends ExprMap {
  var expressions: List[Expr[Any]] = Nil

  override def transform[T](e: Expr[T])(using Type[T])(using q: Quotes): Expr[T] =
    expressions ::= e
    transformChildren(e)

}

def impl[T: Type](expr: Expr[T])(using quotes: Quotes): Expr[T] = {

  val List(es1, es2) = List(MyMap(), MyMap()).map { m =>
    m.transform(expr)
    m.expressions
  }

  assert(es1 == es2)
  assert(es1.map(_.hashCode()) == es2.map(_.hashCode()), s"hash codes not equal:\n${es1.map(_.hashCode())}\n${es2.map(_.hashCode())}")

  expr

}

```

then call the macro in another file with any expression, say `mac(1)`

## Output

```
[error] java.lang.AssertionError: assertion failed: hash codes not equal:
[error] List(290429894, 1544224853)
[error] List(1268450819, 1544224853)
```

## Expectation

the expressions of the two runs are are equal, so the hashcodes should also be equal

--- a/compiler/src/scala/quoted/runtime/impl/ExprImpl.scala
+++ b/compiler/src/scala/quoted/runtime/impl/ExprImpl.scala
@@ -19,5 +19,7 @@ final class ExprImpl(val tree: tpd.Tree, val scope: Scope) extends Expr[Any] {
     case _ => false
   }
 
+  override def hashCode(): Int = tree.hashCode()
+
   override def toString: String = "'{ ... }"
 }

--- a/compiler/src/scala/quoted/runtime/impl/TypeImpl.scala
+++ b/compiler/src/scala/quoted/runtime/impl/TypeImpl.scala
@@ -13,5 +13,7 @@ final class TypeImpl(val typeTree: tpd.Tree, val scope: Scope) extends Type[?] {
     case _ => false
   }
 
+  override def hashCode(): Int = typeTree.hashCode()
+
   override def toString: String = "Type.of[...]"
 }

--- a/tests/pos-macros/i15227a/Macro_1.scala
+++ b/tests/pos-macros/i15227a/Macro_1.scala
@@ -1,0 +1,15 @@
+import scala.quoted.*
+
+inline def mac[T](inline expr: T): T =
+  ${ impl('expr) }
+
+def impl[T: Type](expr: Expr[T])(using Quotes): Expr[T] = {
+  import quotes.reflect.*
+  val expr2 = expr.asTerm.asExpr
+
+  assert(expr == expr2)
+  assert(expr.hashCode() == expr2.hashCode())
+
+  expr
+
+}

--- a/tests/pos-macros/i15227a/Test_2.scala
+++ b/tests/pos-macros/i15227a/Test_2.scala
@@ -1,0 +1,2 @@
+@main def test =
+  mac(1)

--- a/tests/pos-macros/i15227b/Macro_1.scala
+++ b/tests/pos-macros/i15227b/Macro_1.scala
@@ -1,0 +1,27 @@
+import scala.quoted.*
+
+inline def mac[T](inline expr: T): T =
+  ${ impl('expr) }
+
+class MyMap() extends ExprMap {
+  var expressions: List[Expr[Any]] = Nil
+
+  override def transform[T](e: Expr[T])(using Type[T])(using q: Quotes): Expr[T] =
+    expressions ::= e
+    transformChildren(e)
+
+}
+
+def impl[T: Type](expr: Expr[T])(using Quotes): Expr[T] = {
+
+  val List(es1, es2) = List(MyMap(), MyMap()).map { m =>
+    m.transform(expr)
+    m.expressions
+  }
+
+  assert(es1 == es2)
+  assert(es1.map(_.hashCode()) == es2.map(_.hashCode()), s"hash codes not equal:\n${es1.map(_.hashCode())}\n${es2.map(_.hashCode())}")
+
+  expr
+
+}

--- a/tests/pos-macros/i15227b/Test_2.scala
+++ b/tests/pos-macros/i15227b/Test_2.scala
@@ -1,0 +1,2 @@
+@main def test =
+  mac(1)

